Add render blocking post (#12)
siakaramalegos/sia.codes-eleventy: src/posts/render-blocking-resources.md

HTML is technically render blocking because you need it to create the DOM. Without the HTML we would not even have a page to render.

However, HTML is not usually the cause of our problems...

**CSS is render blocking**. The browser needs it before it can create the CSSOM, which blocks all later steps. As soon as the browser encounters a stylesheet `<link>` or `<style>` tag, it must download and parse the contents. Then it must create the CSSOM before the rest of the render can continue. You can see this represented at the triangle point in the [diagram](#critical-render-path). The render tree cannot continue until both the CSSOM and DOM are created.

**JavaScript CAN be render blocking**. It blocks the critical rendering path when it interacts with either the DOM or CSSOM. When the browser encounters a script meant to run synchronously, it will stop DOM creation until the script is finished running:

<figure id="critical-render-path-js">
  <img src="/img/critical_render_path_JS_karamalegos.svg"
    alt="HTML encounters a synchronous script in the head which stops the parser"
    width="595" height="862">

>   <figcaption>Synchronous JavaScript (no async or defer) will block the HTML parser both for download and execution (<a href="#critical-render-path-js">link</a>)</figcaption>

Reviewer: I love the diagrams but it's inaccurate because the CSS will be downloaded before the script executes. This is because of speculative parsing that will look ahead to discover resources, even if the HTML parser is blocked.

It's super tricky to represent that in a diagram though.

https://andydavies.me/blog/2013/10/22/how-the-browser-pre-loader-makes-pages-load-faster/
https://webkit.org/blog/166/optimizing-page-loading-in-web-browser/
https://github.com/whatwg/html/issues/5624

Author: Thanks! I have an idea for that. I'm wondering if that means the CSSOM construction is not blocked? I feel like it should be though.

Reviewer: I have not explored this deeply enough recently to know the answer, sorry.

Author: Thanks again. Just edited the diagram and pushed up all the changes.

Reviewer: That's a good diagram: more accurate without going too deep into speculative parsing.

I think the `<figcaption>` still needs updating though.

Author: Great catch, thanks so much!

Reviewer: Seems like you've updated the other figcaption (and that's a good edit). This one still mentions that it blocks downloads.

Reviewer 2: I presumed a non-async script would block the speculative parser as well as the HTML parser because the script might change the content?

Reviewer: Nope, the speculative parser was created exactly for sync scripts and in particular `document.write()` (see the links I shared). It is called speculative because it could download resources that won't be needed once the real parser as done its job. Bugs aside, the idea works well in practice because we rarely use `document.write` to comment out a section of HTML.

Author: For this one I actually meant the parser is blocked during download and execution of the JavaScript. Not that it blocks other files for download. Is this more clear?

> Synchronous JavaScript (no async or defer) will block the HTML parser during both download and execution of the JavaScript

Reviewer: What about: 
> While synchronous JavaScript is being downloaded and executed, the HTML parser is blocked

</figure>

> Additionally, if CSS has already begun download, the script will stop running to wait until the CSSOM is created before it will continue:

Reviewer: I found the sentence a bit confusing. Here's an attempt:

```suggestion
Additionally, if some CSS appears before a script, the script will not be executed until the CSSOM is created:
```

<blockquote>
  <p>Vodafone improved their LCP by 31%, resulting in an 8% increase in sales, a 15% increase in their lead to visit rate, and an 11% increase in their cart to visit rate.</p>
  <p class="blockquote-source">—<a href="https://wpostats.com/2021/03/18/vodafone-lcp-sales.html">WPO stats</a></p>
</blockquote>

[WPO stats](https://wpostats.com/) lists many cases where improving performance resulted in real business impacts.

## How do I test my website for render-blocking resources?

If you failed this metric in Lighthouse, then you've already discovered one way to test for this. If you're new to Lighthouse, then check out the official [Lighthouse](https://developers.google.com/web/tools/lighthouse) page to get started.

We all have render-blocking resources on our sites (all the CSS!). The problem comes in when it significantly impacts our performance. When this occurs, Lighthouse flags it, and we should do something about it.

Lighthouse candidates for render-blocking resources include both scripts and styles:

> - `<script>` tags in the `<head>` which use do not have at least one of the following attributes: `async`, `defer`, `module`

Reviewer: I think there's an extraneous word:
```suggestion
- `<script>` tags in the `<head>` which do not have at least one of the following attributes: `async`, `defer`, `module`
```
